### PR TITLE
Interpolate UFL expressions

### PR DESCRIPTION
**Pull request by miklos1**
Refactoring in the driver for increased code sharing and loosening of an assertion.

### CODE REVIEW

File: tsfc/impero.py [hunk 1]
```diff
@@ -90,13 +90,13 @@ class Return(Terminal):
     __front__ = ('variable', 'expression')
 
     def __init__(self, variable, expression):
-        assert set(variable.free_indices) == set(expression.free_indices)
+        assert set(variable.free_indices) >= set(expression.free_indices)
```
Comment (wence-) [inline tsfc/impero.py]: OK, so now the lvalue of a return is allowed to have free indices that the rvalue does not.

File: tsfc/impero.py [hunk 2]
```diff
@@ -90,13 +90,13 @@ class Return(Terminal):
...
 
         self.variable = variable
         self.expression = expression
 
     def loop_shape(self, free_indices):
-        return free_indices(self.expression)
+        return free_indices(self.variable)
```
Comment (wence-) [inline tsfc/impero.py]: And so the loop shape changes.

**Comment by wence-** [comment]
This looks fine

File: tsfc/impero.py (hunk 1 above)
Comment (miklos1) [inline tsfc/impero.py]: This comes up, for example, when we interpolate constants onto Functions.

Comment (wence-) [inline tsfc/impero.py]: Is this because we broadcast the constant, or because it's inlined, it doesn't have free indices?

Comment (miklos1) [inline tsfc/impero.py]: A constant is an expression that doesn't depend on the reference point. Yet we have to assign its to each point.

**Pull request merged**